Update ROS params from another node

steph-lheureux opened the pull request:
# Summary
This PR requests the merge of code which provides functionality for nodes to modify each other's [ROS parameters](https://docs.ros.org/en/humble/Concepts/Basic/About-Parameters.html) using [services](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Services/Understanding-ROS2-Services.html). The `RemoteParamsClient` class streamlines setting another node's parameters to a user-friendly abstraction.
# Changes
## Created `RemoteParamsClient` class
### Reasoning
The `RemoteParamsClient` class was created with the primary objective of providing a streamlined process for setting parameters of another node. The class encapsulates underlying complexities of remotely setting parameters, which will become beneficial when we eventually set numerous node's parameters from various nodes across the deck. The abstraction simplifies the process and reduces redundancy which will result in cleaner code and increased ease-of-use.
### Basic Usage
#### Client node
*The client node is the node which makes a request to update another node's parameters*
1. Import the `RemoteParamsClient` class to any node on the deck
	```py
	from .set_remote_params import RemoteParamsClient
	```
2. Create an instance of the class in the `__init__()` function, and pass it `self` (the current node) and the name of the node of whose parameters to set
	```py
	self.__set_params = RemoteParamsClient(self, "other_node_name")
	```
3. In a callback, choose the parameters to set and their new values by calling `update_params()` with arguments of the name of the parameter to set and its new value. This may be called multiple times to set multiple parameters
	```py
	self.__set_params.update_params("param_name", param_val)
	```
4. Once all the parameter values have been instantiated using `update_params()`, send the parameters via service to the other node by calling `send_params()`
	```py
	self.__set_params.send_params()
	```

## Updated naming of nodes
- `input_xbox_one` handles more than just the xbox input now so it has be renamed to `pilot_input`
- `input_keyboard` is unnecessarily wordy and can be shortened to just `keyboard` without losing meaning

Commit:
Changed spacing and fixed spelling

Commit:
Added parameter for throttle curve choice

Commit:
First attempt at remotely editing another node's parameters (this does not work)
Seems to infinitely fail at connecting to the service

Commit:
Basic functionality for a client that updates parameters

Commit:
Subscribes to service publishing parameters of keyboard

Commit:
Abstracts requesting to set another node's parameters via service to a class

Commit:
Moved parameter stuff to use the SetRemotePrarams class

Commit:
Parameter service sets node's parameters

Commit:
General code cleanup
- Updated/added comments
- Fixed spelling errors
- Cleaned up spacing
- Fixed any stylistic errors

Commit:
Renamed nodes and files to be more descriptive
- `input_xbox_one` handles more than just the xbox input now so it has be renamed to `pilot_input`
- `input_keyboard` is unnecessarily wordy and can be shortended to just `keyboard` without losing meaning
- Fixed names in the launch file as well

Commit:
updated names in setup.py

Commit:
Further abstracted setting remote paramaters
The previous versions abstacted the client portion, but not the service. Added another class called SrvRemoteParams which updates
a node's paremeters in one line (when used in conjunction with newly named CliRemoteParams)

steph-lheureux commented on src/seahawk/seahawk_deck/set_remote_params.py:
```diff
@@ -0,0 +1,119 @@
+    """
+    def __init__(self, this_node: Node):
+        """
+        Set up request to setting thi node's parameters from another node
```
Change comment to the following
> Set up request to set this node's parameters from another node

OrionOth commented on src/seahawk/seahawk_deck/keyboard.py:
```diff
@@ -35,60 +35,82 @@
+        # Store the current key
+        self.__cur_key = ""
     
     def __get_key(self) -> str:
```
It doesn't look like this function returns anything anymore, consider changing the signature notation to reflect that?

steph-lheureux replied:
Thanks, it returned something at some point and I forgot to remove the type hint. Updated to
```py
def __get_key(self):
```

OrionOth commented on src/seahawk/seahawk_deck/keyboard.py:
```diff
@@ -35,60 +35,82 @@
+        """
+        Publishes the current key to 'keystroke' topic
+
+        Throws 'KeyboardInterrupt' if user enters ctrl-c
```
I don't believe this comment is accurate anymore; the `KeyboardInterrupt` is thrown above in the `callback` function.

OrionOth commented on src/seahawk/seahawk_deck/keyboard.py:
```diff
@@ -35,60 +35,82 @@
         self.__key_pub = self.create_publisher(String, "keystroke", 10)
 
+        # Set up client to remotely set parameters on 'pilot_input' node using a service
+        self.__set_params_input_xbox_one = CliRemoteParams(self, "pilot_input")
```
Given renaming elsewhere, I think you should rename `__set_params_input_xbox_one` and related to something more like `__set_params_pilot_input` for better consistency.

steph-lheureux replied:
Fixed
```py
# Send all params to pilot_input node
        self.__set_params_pilot_input.send_params()
```

steph-lheureux replied:
![Screenshot from 2024-01-15 18-23-13](https://github.com/CabrilloRoboticsClub/cabrillo_rov_2023/assets/96219204/2a2aa9f6-a9b7-45a4-aad9-9911c8796a0b)

OrionOth commented on src/seahawk/seahawk_deck/keyboard.py:
```diff
@@ -35,60 +35,82 @@
+        """
+        Updates parameters for 'throttle_curve_choice'
+        """
+        # If the user pressed '0', '1', or '2' send that information to the __set_params_input_xbox_one
+        if self.__cur_key in ["0", "1", "2"]:
+            self.__set_params_input_xbox_one.update_params("throttle_curve_choice", self.__cur_key)
```
This is a very unergonomic method of keybinding. If we wanted to change these to, say, letters instead, then the letter would be sent. It's better to abstract this I'd say. Consider using a dict, or the index in the array, and sending numbers 0-3 instead of characters "0"-"3", or using id-like names (`"linear"`, `"gentle"`, `"deadzone-50"`...).

steph-lheureux replied:
[wont-fix] This will be updated in the gui implementation of the keyboard, no use fixing something which will be re-written #186

OrionOth commented on src/seahawk/seahawk_deck/pilot_input.py:
```diff
@@ -77,21 +79,30 @@ def reset(self):
         self.__claw_pub = self.create_publisher(Bool, "claw_state", 10)
-        
+
+        # Create and store parameter which determines which throttle curve
+        # the pilot wants to use named 'throttle_curve_choice'. Defaults to '0'
+        self.declare_parameter("throttle_curve_choice", "0")
```
See my other comment on `keyboard.py` about unergonomic keybinding

steph-lheureux replied:
See other comment above

OrionOth reviewed (changes requested):
I like the node name changes.

I am very much not a fan of the confusing names `SrvRemoteParams` and `CliRemoteParams`. Maybe `RemoteParamsService` and `RemoteParamsClient`? Also, perhaps consider changing some of the wording/comments around the "service" to specify that it isn't so much a service node in itself but rather is a utility class to add a service to a node. `RemoteParamsServiceWrapper`? I do like the way you are implementing it, though.

TLDR; my gripes are around naming conventions mostly

OrionOth commented on src/seahawk/seahawk_deck/set_remote_params.py:
```diff
@@ -0,0 +1,119 @@
+        return future.result()                      # Return if the parameter is set correctly
+    
+
+class SrvRemoteParams():
+    """
+    Service node. Abstracts requesting setting node's parameters from another node via service to a class.
+    """
+    def __init__(self, this_node: Node):
+        """
+        Set up request to setting thi node's parameters from another node
+    
+        Args:
+            this_node: Name of the node whose parameters should be set. Service node
+        """
+        self.__this_node = this_node
+        # Call parameter callback
+        self.__this_node.add_on_set_parameters_callback(self.__set_params)
+    
+    def __set_params(self, params: list[Parameter]) -> SetParametersResult:
+        """
+        Callback for parameter update. Updates all local parameters to have values set by the service
+
+        Args:
+            params: List of updated parameters
+
+        Returns:
+            SetParametersResult() which lets ROS2 know if the parameters were set correctly or not
+        """
+        try:
+            # Try to set parameters to updated values
+            self.__this_node.set_parameters(params)
+        except:
+            return SetParametersResult(successful=False)
+        return SetParametersResult(successful=True)
```
This is effectively re-implementing what ROS is already doing, setting the parameters. This leads to unpredictable behavior, likely recursive.

https://docs.ros.org/en/rolling/Concepts/Basic/About-Parameters.html#parameter-callbacks
> A ROS 2 node can register three different types of callbacks to be informed when changes are happening to parameters. All three of the callbacks are optional.
> ...
> The second is known as a “set parameter” callback, and can be set by calling `add_on_set_parameters_callback` from the node API. The callback is passed a list of immutable `Parameter` objects, and returns an `rcl_interfaces/msg/SetParametersResult`. The main purpose of this callback is to give the user the ability to inspect the upcoming change to the parameter and explicitly reject the change.
> ...

here's an example of what `add_on_set_parameters_callback` is supposed to be doing: https://github.com/ros2/demos/blob/rolling/demo_nodes_cpp/src/parameters/set_parameters_callback.cpp#L70

steph-lheureux replied:
Nice catch, I removed the entire `SrvRemoteParams` class and it still works

OrionOth reviewed (changes requested):
`SrvRemoteParams` is completely unnecessary here.

Commit:
Fixed minor naming and comment mistakes

Commit:
Changed names to make them more clear

Commit:
Removed unnecessary SrvRemoteParams class
ROS2 already handles setting parameters so there is no reason to have this class. The code works without it and including this class may cause unexpected behaviour

Commit:
Removed references to SrvRemoteParams

OrionOth committed:
Revert changes related to keyboard node

OrionOth reviewed (approved):
LGTM! Don't forget to make another PR with the commit from `git revert 757e2af` in order to get the keyboard changes too.

OrionOth commented on src/seahawk/seahawk_deck/set_remote_params.py:
```diff
@@ -0,0 +1,85 @@
+            this_node: Node which makes the request to set another node's parameters. Client node
+            other_node_name: Name of the node whose parameters should be set. Service node
+        """
+        # Create service name. Services to set parameters must be named as '/other_node_name/set_parameters'
+        srv_name = "/" + other_node_name + "/set_parameters"
```
I think it would be somewhat "more correct" to not have the prefixing `/` as there is a sort of namespace hierarchy in ROS.

steph-lheureux replied:
Services are always in the form `/node_name/srv_name`
![services](https://github.com/CabrilloRoboticsClub/cabrillo_rov_2023/assets/96219204/8597de08-f854-4cb4-b864-b98d0c10bcf3)

OrionOth replied:
Yes, but think of it like absolute paths versus relative paths. ROS will prefix it with whatever it needs.
At least, that's my interpretation of how it works.

Commit:
Reverted keyboard node to state on the main branch
The keyboard node will likely be moved to the GUI so setting the parameters from this node was
unnessicary. The only new changes are the name of the node which was changed from 'input_keyboard'
to 'keyboard'

Commit:
Merge branch 'remote_ros_params' of github.com:CabrilloRoboticsClub/cabrillo_rov_2023 into remote_ros_params

Commit:
Updated names in launch files

Commit:
Renamed file to be shorter

Commit:
Revert "Updated names in launch files"

This reverts commit 81a9ae501c557cb49fa9a9f3f6cf588509d2fb39.

Commit:
Changed all occurances of 'input_keyboard' to 'keyboard'

steph-lheureux committed:
Revert "Revert changes related to keyboard node"

This reverts commit 757e2af1741bac4921d269fab9112e228ca0e610.

mike-matera commented on src/seahawk/seahawk_deck/set_remote_params.py:
```diff
@@ -0,0 +1,85 @@
+        """
+        self.__param_list.append(Parameter(name=param_name, value=param_value).to_parameter_msg())
+	
+    def send_params(self) -> (Any | None):
```
This is great code Steph. It makes it way easier to grab and hold onto a remote parameter that you're likey to change often. Here's something to consider. It looks like users of this have to:

```python 
p.update_params('foo', 'bar')
p.update_params('thing', 'bak')
p.send_params()
```

If you liked, you could combine the calls by making `send_params` take a dictionary:

```python
p.send_params({
  'foo': 'bar', 
  'thing': 'bak',
})
```

Just something to consider.

OrionOth replied:
This history is going to be rewritten as we discussed a few minutes ago, so this review will be lost. Sorry! Normally I wouldn't ever rewrite history on an active PR but this is emergency mode I think.

steph-lheureux replied:
Thanks!! In the current use cases I can think of, we are only going to be updating one parameter at a time, what would be an instance where we would want to update multiple parameters? I could not think of a case when writing the code, but knew maybe it could be a possibility and thus allowed the user to call `update_params()` multiple times. In my opinion it is not necessary to optimize for an unlikely case but let me know if there is something I am not considering

OrionOth replied:
The main issue with parameters is there are only a few fixed types. If you wanted to use the thrust matrix as a parameter, for example, it would have to be treated as an array with separate parameter(s) to describe how many columns/rows it has.

liamgilligan replied:
Can't help but give a list comprehension for dealing with that:

`self.__req.parameters = [Parameter(name=param_name, value=param_value).to_parameter_msg() for param_name, param_value in self.__param_list.items()]`

liamgilligan replied:
That isn't an issue @OrionOth, at least in regards to the motor configuration matrix. The parameter that handles variable center of mass is a three element array that indicates the offset of the center of mass in (x, y, z) (see 694f221).

OrionOth commented:
This force push was because all of those commits were undesired by Steph.

mike-matera reviewed (approved):
Looks good!

liamgilligan commented on src/seahawk/seahawk_deck/pilot_input.py:
```diff
@@ -77,21 +76,27 @@ def reset(self):
+
+        # Create and store parameter which determines which throttle curve
+        # the pilot wants to use named 'throttle_curve_choice'. Defaults to '0'
+        self.declare_parameter("throttle_curve_choice", "0")
```
Suggestion (strong): The parameter is created but not stored. You should create a member variable (such as `self.throttle_curve_choice`) to hold the value of the parameter so that you don't have to call `self.get_parameter("throttle_curve_choice").value` to get the value every time.

steph-lheureux replied:
```py
# Create and store parameter which determines which throttle curve
# the pilot wants to use named 'throttle_curve_choice'. Defaults to '0'
self.declare_parameter("throttle_curve_choice", "0")
self.__throttle_curve_choice = self.get_parameter("throttle_curve_choice").value
```
Added member variable

OrionOth replied:
This isn't set anywhere else.

liamgilligan commented on src/seahawk/seahawk_deck/set_remote_params.py:
```diff
@@ -0,0 +1,85 @@
+
+class RemoteParamsClient():
+    """
+    Abstracts requesting to set another node's parameters via service to a class. 
```
Nit: change "via service" to "via a service call".

liamgilligan replied:
There are multiple instances of this, this nitpick applies to all.

liamgilligan commented on src/seahawk/seahawk_deck/set_remote_params.py:
```diff
@@ -0,0 +1,85 @@
+    
+        Args:
+            this_node: Node which makes the request to set another node's parameters. Client node
+            other_node_name: Name of the node whose parameters should be set. Service node
```
Nit: Change "Service node" to "Parameter node", as this class only changes parameters and does not handle service calls broadly.

liamgilligan commented on src/seahawk/seahawk_deck/set_remote_params.py:
```diff
@@ -0,0 +1,85 @@
+        Sends updated parameter list to the service
+
+        Returns:
+            The result set by the task, None if no result was set
```
Nit: Confusing `Returns:` description in the docstring, consider elaborating.

liamgilligan reviewed (approved):
Overall, looks super good. I could only find one real area to change, which is between an Issue and a Suggestion (I marked it as a strong suggestion). Otherwise, I could only find nitpicks. A lot of the stylistic changes you made are super good to see, and the added functionality is awesome.

I'm marking as approved but I would *strongly* prefer you to consider some of the changes I recommended, however code is good as is IMO. Specifically, consider creating a member variable for storing the value of the `throttle_curve_choice` parameter in the `pilot_input` node. 

Great work!

steph-lheureux committed:
Fixed nit picks left in pr review
- Nit: change "via service" to "via a service call".
- Nit: Change "Service node" to "Parameter node", as this class only changes parameters and does not handle service calls broadly.
- Nit: Confusing Returns [of send_params()]: description in the docstring, consider elaborating.
- Changed name of class to SetRemoteParams as there are no longer both "client" and "sevice" classes

steph-lheureux committed:
Resolved stromg suggestion regarding storing a member variable of the parameter value
Suggestion (strong): The parameter is created but not stored. You should create a member variable (such as self.throttle_curve_choice)
to hold the value of the parameter so that you don't have to call self.get_parameter("throttle_curve_choice").value to get the value every time.

steph-lheureux committed:
Updated outdated comment
